From 1e537f8d0916f1c90a71fb5e1cda3b51dc1aef04 Mon Sep 17 00:00:00 2001
From: Andrey Grodzovsky <Andrey.Grodzovsky@amd.com>
Date: Fri, 12 Aug 2016 14:40:40 -0400
Subject: [PATCH 0699/1722] drm/amd/dal: Fix DAL crash on boot with 4.7 kernel
 (v2)

Double call to unreference a connector was leading to
connector's unregistration and destruction.

v2: Rework atomic_duplicate_state DAL callback
for connector so a refrence will be aquired on first state
allocation. Use default implementation of atomic_destroy_state
instead of DAL specific which currently doesn't
add funcationality.

Change-Id: I9506d4d2fb120410841478e7e59490d9242fdc00
Reviewed-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Andrey Grodzovsky <Andrey.Grodzovsky@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 .../gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c  |  2 +-
 drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c  | 20 ++++----------------
 drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.h  |  4 +---
 3 files changed, 6 insertions(+), 20 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
index f7844b6..d148717 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_mst_types.c
@@ -159,7 +159,7 @@ static const struct drm_connector_funcs dm_dp_mst_connector_funcs = {
 	.reset = amdgpu_dm_connector_funcs_reset,
 	.set_property = drm_atomic_helper_connector_set_property,
 	.atomic_duplicate_state = amdgpu_dm_connector_atomic_duplicate_state,
-	.atomic_destroy_state = amdgpu_dm_connector_atomic_destroy_state,
+	.atomic_destroy_state = drm_atomic_helper_connector_destroy_state,
 	.atomic_set_property = amdgpu_dm_connector_atomic_set_property
 };
 
diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
index f3e69d3..f013176 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.c
@@ -1362,29 +1362,17 @@ struct drm_connector_state *amdgpu_dm_connector_atomic_duplicate_state(
 		to_dm_connector_state(connector->state);
 
 	struct dm_connector_state *new_state =
-		kzalloc(sizeof(*new_state), GFP_KERNEL);
+                        kmemdup(state, sizeof(*state), GFP_KERNEL);
 
 	if (new_state) {
-		*new_state = *state;
-
+                __drm_atomic_helper_connector_duplicate_state(connector,
+                                                                      &new_state->base);
 		return &new_state->base;
 	}
 
 	return NULL;
 }
 
-void amdgpu_dm_connector_atomic_destroy_state(
-	struct drm_connector *connector,
-	struct drm_connector_state *state)
-{
-	struct dm_connector_state *dm_state =
-		to_dm_connector_state(state);
-
-	__drm_atomic_helper_connector_destroy_state(connector, state);
-
-	kfree(dm_state);
-}
-
 static const struct drm_connector_funcs amdgpu_dm_connector_funcs = {
 	.dpms = drm_atomic_helper_connector_dpms,
 	.reset = amdgpu_dm_connector_funcs_reset,
@@ -1393,7 +1381,7 @@ static const struct drm_connector_funcs amdgpu_dm_connector_funcs = {
 	.set_property = drm_atomic_helper_connector_set_property,
 	.destroy = amdgpu_dm_connector_destroy,
 	.atomic_duplicate_state = amdgpu_dm_connector_atomic_duplicate_state,
-	.atomic_destroy_state = amdgpu_dm_connector_atomic_destroy_state,
+	.atomic_destroy_state = drm_atomic_helper_connector_destroy_state,
 	.atomic_set_property = amdgpu_dm_connector_atomic_set_property
 };
 
diff --git a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.h b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.h
index 15c22fe..4f7bd3b 100644
--- a/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.h
+++ b/drivers/gpu/drm/amd/dal/amdgpu_dm/amdgpu_dm_types.h
@@ -67,9 +67,7 @@ int dm_create_validation_set_for_target(
 void amdgpu_dm_connector_funcs_reset(struct drm_connector *connector);
 struct drm_connector_state *amdgpu_dm_connector_atomic_duplicate_state(
 	struct drm_connector *connector);
-void amdgpu_dm_connector_atomic_destroy_state(
-	struct drm_connector *connector,
-	struct drm_connector_state *state);
+
 int amdgpu_dm_connector_atomic_set_property(
 	struct drm_connector *connector,
 	struct drm_connector_state *state,
-- 
2.7.4

